libraries/c_sdk/standard/mqtt/src/iot_mqtt_subscription_container.c

```diff
@@ -0,0 +1,357 @@
+    }
+
+    while( ( nameIndex < topicNameLength ) && ( filterIndex < topicFilterLength ) )
+    {
+        /* Check if the character in the topic name matches the corresponding
+         * character in the topic filter string. */
+        if( pTopicName[ nameIndex ] == pTopicFilter[ filterIndex ] )
+        {
+            /* Handle special corner cases as documented by the MQTT protocol spec. */
+
+            /* Filter "sport/#" also matches "sport" since # includes the parent level. */
+            if( nameIndex == topicNameLength - 1 )
+            {
+                if( filterIndex == topicFilterLength - 3 )
+                {
+                    if( pTopicFilter[ filterIndex + 1 ] == '/' )
+                    {
+                        if( pTopicFilter[ filterIndex + 2 ] == '#' )
+                        {
+                            IOT_SET_AND_GOTO_CLEANUP( true );
+                        }
+                    }
+                }
+            }
+
+            /* Filter "sport/+" also matches the "sport/" but not "sport". */
+            if( nameIndex == topicNameLength - 1 )
+            {
+                if( filterIndex == topicFilterLength - 2 )
+                {
+                    if( pTopicFilter[ filterIndex + 1 ] == '+' )
+                    {
+                        IOT_SET_AND_GOTO_CLEANUP( true );
+                    }
+                }
+            }
+        }
+        else
+        {
+            /* Check for wildcards. */
+            if( pTopicFilter[ filterIndex ] == '+' )
+            {
+                /* Move topic name index to the end of the current level.
+                 * This is identified by '/'. */
+                while( nameIndex < topicNameLength && pTopicName[ nameIndex ] != '/' )
+                {
+                    nameIndex++;
+                }
+
+                /* Increment filter index to skip '/'. */
+                filterIndex++;
+                continue;
+            }
+            else if( pTopicFilter[ filterIndex ] == '#' )
+            {
+                /* Subsequent characters don't need to be checked if the for the
+                 * multi-level wildcard. */
+                IOT_SET_AND_GOTO_CLEANUP( true );
+            }
+            else
+            {
+                /* Any character mismatch other than '+' or '#' means the topic
+                 * name does not match the topic filter. */
+                IOT_SET_AND_GOTO_CLEANUP( false );
+            }
+        }
+
+        /* Increment indexes. */
+        nameIndex++;
+        filterIndex++;
+    }
+
+    /* If the end of both strings has been reached, they match. */
+    if( ( nameIndex == topicNameLength ) && ( filterIndex == topicFilterLength ) )
+    {
+        IOT_SET_AND_GOTO_CLEANUP( true );
+    }
+
+    IOT_FUNCTION_EXIT_NO_CLEANUP();
+}
+
+/*-----------------------------------------------------------*/
+
+int8_t IotMqtt_InsertSubscription( _mqttSubscription_t * pSubscriptionArray )
```

Comment:
Isn't this getting the freeIndex? name says different action

Reply:
Addressed